ofono/plugins/cellinfo-netmon.c

```diff
@@ -0,0 +1,486 @@
+	return G_SOURCE_REMOVE;
+}
+
+static void cellinfo_netmon_request_update_destroy(gpointer data)
+{
+	struct cellinfo_netmon_update_cbd *cbd = data;
+	struct cellinfo_netmon_data *nm = cbd->nm;
+
+	cell_info_control_drop_requests(nm->ctl, cbd);
+	ofono_cell_info_remove_handler(cbd->info, cbd->event_id);
+	ofono_cell_info_unref(cbd->info);
+	g_free(cbd);
+}
+
+static void cellinfo_netmon_request_update(struct ofono_netmon *netmon,
+		ofono_netmon_cb_t cb, void *data)
+{
+	struct cellinfo_netmon_data *nm = cellinfo_netmon_get_data(netmon);
+	struct ofono_cell_info *info = nm->ctl->info;
+	struct cellinfo_netmon_update_cbd *cbd =
+		g_new(struct cellinfo_netmon_update_cbd, 1);
+
+	cbd->cb = cb;
+	cbd->data = data;
+	cbd->nm = nm;
+	cbd->info = ofono_cell_info_ref(info);
+	cbd->event_id = ofono_cell_info_add_change_handler(info,
+				cellinfo_netmon_request_update_event, cbd);
+
+	/* Temporarily enable updates and wait */
+	DBG("%s waiting for update", nm->ctl->path);
+	cell_info_control_set_update_interval(nm->ctl, cbd,
+				NETMON_UPDATE_INTERVAL_MS);
+	cell_info_control_set_enabled(nm->ctl, cbd, TRUE);
+
+	/* Use shorter timeout if we already have something */
+	nm->update_id = g_timeout_add_full(G_PRIORITY_DEFAULT_IDLE,
+				cellinfo_netmon_have_registered_cells(info) ?
+				NETMON_UPDATE_SHORT_TIMEOUT_MS :
+				NETMON_UPDATE_LONG_TIMEOUT_MS,
+				cellinfo_netmon_request_update_timeout,
+				cbd, cellinfo_netmon_request_update_destroy);
+}
+
+static void cellinfo_netmon_enable_periodic_update(struct ofono_netmon *netmon,
+		unsigned int enable, unsigned int period_sec,
+		ofono_netmon_cb_t cb, void *data)
+{
+	struct cellinfo_netmon_data *nm = cellinfo_netmon_get_data(netmon);
+	CellInfoControl *ctl = nm->ctl;
+
+	if (ctl) {
+		const int ms = period_sec * 1000;
+
+		if (enable) {
+			cell_info_control_set_update_interval(ctl, nm, ms);
+			cell_info_control_set_enabled(ctl, nm, TRUE);
+		} else {
+			cell_info_control_set_enabled(ctl, nm, FALSE);
+			cell_info_control_set_update_interval(ctl, nm, ms);
+		}
+	}
+
+	CALLBACK_WITH_SUCCESS(cb, data);
+}
+
+static gboolean cellinfo_netmon_register(gpointer user_data)
+{
+	struct cellinfo_netmon_data *nm = user_data;
+
+	nm->register_id = 0;
+	ofono_netmon_register(nm->netmon);
+
+	return G_SOURCE_REMOVE;
+}
+
+static int cellinfo_netmon_probe(struct ofono_netmon *netmon,
+		unsigned int vendor, void *modem)
+{
+	const char *path = ofono_modem_get_path(modem);
+	struct cellinfo_netmon_data *nm =
+		g_new0(struct cellinfo_netmon_data, 1);
+
+	nm->netmon = netmon;
+	nm->ctl = cell_info_control_get(path);
+
+	ofono_netmon_set_data(netmon, nm);
+	nm->register_id = g_idle_add(cellinfo_netmon_register, nm);
+	DBG("%s", path);
+
+	return 0;
+}
+
+static void cellinfo_netmon_remove(struct ofono_netmon *netmon)
+{
+	struct cellinfo_netmon_data *nm = cellinfo_netmon_get_data(netmon);
+
+	DBG("%s", nm->ctl ? nm->ctl->path : "?");
+	ofono_netmon_set_data(netmon, NULL);
+
+	if (nm->update_id) {
```

Comment:
Again kernel style, as ihnerited by ofono, suggests not using curlies here, your choice.

Reply:
Yeah, I know, but being bitten by a macro splitting what looks like a single statement into multiple statements, I would rather surround even a single statement with curlies (except for maybe things like `i++` which pose no such danger). It's a kind of a psychological trauma which is hard to overcome)) And it might be one of the reasons why I like to make APIs null tolerant - if `g_source_remove()` were ignoring zeros, then these particular checks wouldn't be necessary in the first place.